ARCs/arc-0035.md

```diff
@@ -0,0 +1,119 @@
+
+### Requirements
+At a high-level, offline wallet backup protocol has the following requirements:
+
+ * Wallet applications should allow backing up and storing multiple accounts at the same time.
+Account information should be encrypted with a user defined secret key, utilizing NaCl SecretBox method (audited and endorsed by Algorand).
+ 
+ * Encrypted final string should be easily copyable to be stored digitally.
+When importing, wallet applications should be able to detect already imported accounts and gracefully ignore them.
+
+### Format
+
+Before encryption, account information should be converted to the following JSON format:
+
+```
+{
+  "device_id": "UNIQUE IDENTIFIER FOR DEVICE (OPTIONAL)",
+  "provider_name": "PROVIDER NAME (OPTIONAL, i.e. Pera Wallet)",
+  "accounts": [
+    {
+      "address": "ACCOUNT PUBLIC ADDRESS (REQUIRED)"
+      "name": "USER DEFINED ACCOUNT NAME (OPTIONAL)",
+      "account_type": "TYPE OF ACCOUNT: single, multisig, watch, contact, ledger (REQUIRED)",
+    "private_key": "ACCOUNT PRIVATE KEY AS BASE64 ENCODING OF 64 BYTE PRIVATE KEY as encoded by NACL/libsodium (NOT PASSPHRASE, REQUIRED for user-owned accounts, can be omitted in case of watch, contact, ledger accounts)"
+      "metadata": "ANY ADDITIONAL CONTENT (OPTIONAL)",
+    },
+    ...
+  ],
+}
+```
+
+*Clients must accept additional fields in the JSON document.*
+
+Here is an example with a single account:
+
+```
+{
+  "device_id": "2498232091970170817",
+  "provider_name": "Pera Wallet",
+  "accounts": [
+    {
+      "address": "ELWRE6HZ7KIUT46EQ6PBISGD3ND6QSCBVWICYR2QR2Y7LOBRZRCAIKLWDE"
+      "name": "My NFT Account",
+      "account_type": "single",
+      "private_key": "w0HG2VH7tAYz9PD4SYX0flC4CKh1OONCB6U5bP7cXGci7RJ4+fqRSfPEh54USMPbR+hIQa2QLEdQjrH1uDHMRA=="
+    }
+  ],
+}
+```
+
+### Encryption
+
+Once the input JSON is ready as specified above, it needs to be encrypted. Even if it is assumed that the user is going to store this information in a secure location, copy-pasting it without encryption is not secure since multiple applications can access the clipboard. 
+
+The information needs to be encrypted using a very long passphrase. 12 word mnemonic will be used as the key. 12-word mnemonic is secure and it will not create confusion with the 24-word mnemonics that are used to encrypt accounts.
+
+The wallet applications should not allow users to copy the 12-word mnemonic nor allow taking screenshots. The users should note it visually. 
+
+The encryption should be made as follows:
+
+1. The wallet generates a random 16-byte string S (using a cryptographically secure random number generator)
+2. The wallet derives a 32-byte key: `key = HMAC-SHA256(key="Algorand export 1.0", input=S)`
```

Comment:
Unfortunately [libsodiumjs](https://github.com/jedisct1/libsodium.js/) does not have exposed the functions: crypto_auth_hmacsha256_init / crypto_auth_hmacsha256_update / crypto_auth_hmacsha256_final. Only has wrapped crypto_auth_hmacsha256 which only supports a fixed key of 32 bytes [here](https://github.com/jedisct1/libsodium/blob/6d566070b48efd2fa099bbe9822914455150aba9/src/libsodium/crypto_auth/hmacsha256/auth_hmacsha256.c#L101).

What can we do instead? A fixed text of 32bytes, in stead of  "`Algorand export 1.0`"? Or add padding to complete the 32bytes? Any other suggestion?

Regards

Reply:
@OEA, what are you using here?  As @elranu states, it may be good to change this text.
cc @fergalwalsh

Reply:
I used Go's  `crypto/hmac` and `crypto/sha256` packages here.

This is the example code of generating 32-byte key in Go:

```
func GenerateCipherKey(input []byte) ([]byte) {
	cipher := hmac.New(sha256.New, []byte("Algorand export 1.0"))
	cipher.Write(input)
	return cipher.Sum(nil) 
}
```

Reply:
Ok, so libsodium is not mandatory

Reply:
Ok, and in the case of step 3, using `crypto_secretbox_easy` how do we assign and store the nonce value?
Because nonce can't be null, I see this example in [libsodiumjs](https://github.com/jedisct1/libsodium.js/#api) doc, they concat the nonce value with the ciphertext:

```
let key = sodium.from_hex('724b092810ec86d7e35c9d067702b31ef90bc43a7b598626749914d6a3e033ed');

function encrypt_and_prepend_nonce(message) {
    let nonce = sodium.randombytes_buf(sodium.crypto_secretbox_NONCEBYTES);
    return nonce.concat(sodium.crypto_secretbox_easy(message, nonce, key));
}

function decrypt_after_extracting_nonce(nonce_and_ciphertext) {
    if (nonce_and_ciphertext.length < sodium.crypto_secretbox_NONCEBYTES + sodium.crypto_secretbox_MACBYTES) {
        throw "Short message";
    }
    let nonce = nonce_and_ciphertext.slice(0, sodium.crypto_secretbox_NONCEBYTES),
        ciphertext = nonce_and_ciphertext.slice(sodium.crypto_secretbox_NONCEBYTES);
    return sodium.crypto_secretbox_open_easy(ciphertext, nonce, key);
}
```

Is this the right thing to do, or is another strategy being considered?

Reply:
@yuli-ferna As far as I see in this example, it looks like right. Because the nonce value always randomly generated. When you decrypt it, you can slice the nonce and pass it to necessary function.

Reply:
Great! Thank you! In this case, @yigitguler don't you think we should specify how to save the nonce in the doc?

Reply:
Yes, I think it would be great. Could you please add a suggestion so that we can add?

Thank you.

Reply:
Why not use an actual KDF? (e.g HKDF)